File: bench/benches/strcat.py
import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "lib"))
from bench import bench

HMOD = 1000000007
N = 4000

# build an N-char string by repeated single-char concatenation, then hash it.
def work():
    s = ""
    for i in range(N):
        s += chr(48 + i % 10)
    h = 0
    for ch in s:
        h = (h * 31 + ord(ch)) % HMOD
    return h

bench("strcat", work)

File: bench/benches/strcat.java
// build an N-char string by repeated single-char concatenation, then hash it.
class Main {
    static final long HMOD = 1000000007L;
    static final int N = 4000;

    static long work() {
        String s = "";
        for (int i = 0; i < N; i++) {
            s += (char) (48 + i % 10);
        }
        long h = 0;
        for (int i = 0; i < s.length(); i++) {
            h = (h * 31 + s.charAt(i)) % HMOD;
        }
        return h;
    }

    public static void main(String[] a) {
        Bench.bench("strcat", Main::work);
    }
}
